File: lib/web/js/response.js
import * as ops from 'response.go'
import { Blob } from 'web:blob'

const kResponseId = Symbol('responseId')
const kHeaders = Symbol('headers')
const kBodyUsed = Symbol('bodyUsed')

const encoder = new TextEncoder()
const decoder = new TextDecoder()

const VALID_REDIRECT_STATUSES = new Set([301, 302, 303, 307, 308])

function toHeaders(init) {
    if (init instanceof Headers) {
        return init
    }
    return new Headers(init ?? {})
}

function headersToRecord(headers) {
    const bag = toHeaders(headers)
    const record = {}
    bag.forEach((value, key) => {
        record[key] = value
    })
    return record
}

function recordToHeaders(record) {
    return record instanceof Headers ? record : new Headers(record ?? {})
}

function collectBlobPartIds(blob) {
    const ids = []

    const walk = (target) => {
        if (!(target instanceof Blob)) {
            return
        }

        const parts = target._parts ?? []
        for (const part of parts) {
            if (part instanceof Blob) {
                walk(part)
            } else if (part && typeof part._id === 'string') {
                ids.push(part._id)
            }
        }
    }

    walk(blob)
    return ids
}

function normalizeBody(body, headers) {
    if (body === null || body === undefined) {
        return { kind: 'none' }
    }

    if (body instanceof Blob) {
        if (body.type && !headers.has('content-type')) {
            headers.set('content-type', body.type)
        }
        return {
            kind: 'blob',
            blobParts: collectBlobPartIds(body),
            contentType: body.type || null,
        }
    }

    if (body instanceof ArrayBuffer) {
        return { kind: 'bytes', bytes: new Uint8Array(body) }
    }

    if (ArrayBuffer.isView(body)) {
        return {
            kind: 'bytes',
            bytes: new Uint8Array(body.buffer, body.byteOffset, body.byteLength),
        }
    }

    if (typeof body === 'string') {
        if (!headers.has('content-type')) {
            headers.set('content-type', 'text/plain;charset=UTF-8')
        }
        return {
            kind: 'bytes',
            bytes: encoder.encode(body),
            contentType: headers.get('content-type'),
        }
    }

    if (body instanceof URLSearchParams) {
        if (!headers.has('content-type')) {
            headers.set('content-type', 'application/x-www-form-urlencoded;charset=UTF-8')
        }
        return {
            kind: 'bytes',
            bytes: encoder.encode(body.toString()),
            contentType: headers.get('content-type'),
        }
    }

    if (body instanceof Uint8Array) {
        return { kind: 'bytes', bytes: new Uint8Array(body) }
    }

    throw new TypeError('Unsupported Response body type')
}

function createSnapshot(metadata, headers, bodyDescriptor) {
    const bag = toHeaders(headers)
    return {
        status: metadata.status,
        statusText: metadata.statusText,
        headers: headersToRecord(bag),
        url: metadata.url,
        type: metadata.type,
        bodyUsed: false,
        hasBody: bodyDescriptor.kind !== 'none',
        contentType: bag.get('content-type') ?? bodyDescriptor.contentType ?? '',
    }
}

function createResponseFromSnapshot(id, snapshot) {
    const instance = Object.create(Response.prototype)
    instance[kResponseId] = id
    instance._status = snapshot.status
    instance._statusText = snapshot.statusText || ''
    instance._url = snapshot.url || ''
    instance._type = snapshot.type || 'basic'
    instance[kHeaders] = recordToHeaders(snapshot.headers)
    instance[kBodyUsed] = Boolean(snapshot.bodyUsed)
    instance._contentType = snapshot.contentType ?? ''
    return instance
}

async function consumeBodyBytes(response) {
    if (response[kBodyUsed]) {
        throw new TypeError('Body has already been used for this Response')
    }

    response[kBodyUsed] = true
    const result = ops.op_response_consume_body(response[kResponseId])

    if (!result) {
        return new Uint8Array()
    }

    if (result instanceof Uint8Array) {
        return result
    }

    if (result instanceof ArrayBuffer) {
        return new Uint8Array(result)
    }

    if (ArrayBuffer.isView(result)) {
        return new Uint8Array(result.buffer, result.byteOffset, result.byteLength)
    }

    return Uint8Array.from(result)
}

class Response {
    constructor(body = null, init = {}) {
        const status = init.status ?? 200
        if (status < 200 || status > 599) {
            throw new RangeError('Response status must be between 200 and 599')
        }

        const statusText = init.statusText ?? ''
        const headers = toHeaders(init.headers)

        const bodyDescriptor = normalizeBody(body, headers)
        if ((status === 204 || status === 205 || status === 304) && bodyDescriptor.kind !== 'none') {
            throw new TypeError('Body not allowed for status 204, 205, or 304')
        }

        const metadata = {
            status,
            statusText,
            headers: headersToRecord(headers),
            url: init.url ?? '',
            type: init.type ?? 'basic',
            bodyKind: bodyDescriptor.kind,
            blobParts: bodyDescriptor.blobParts ?? [],
            contentType: headers.get('content-type') ?? bodyDescriptor.contentType ?? '',
        }

        const responseId = bodyDescriptor.kind === 'bytes'
            ? ops.op_response_create(metadata, bodyDescriptor.bytes ?? new Uint8Array())
            : ops.op_response_create(metadata)

        const snapshot = createSnapshot(metadata, headers, bodyDescriptor)
        return createResponseFromSnapshot(responseId, snapshot)
    }

    get status() {
        return this._status
    }

    get ok() {
        return this._status >= 200 && this._status < 300
    }

    get statusText() {
        return this._statusText
    }

    get type() {
        return this._type
    }

    get url() {
        return this._url
    }

    get headers() {
        return this[kHeaders]
    }

    get bodyUsed() {
        return this[kBodyUsed]
    }

    async arrayBuffer() {
        const bytes = await consumeBodyBytes(this)
        return bytes.buffer.slice(bytes.byteOffset, bytes.byteOffset + bytes.byteLength)
    }

    async text() {
        const buffer = await this.arrayBuffer()
        return decoder.decode(buffer)
    }

    async json() {
        const text = await this.text()
        return JSON.parse(text)
    }

    async blob() {
        const bytes = await consumeBodyBytes(this)
        return new Blob([bytes], { type: this._contentType ?? this.headers.get('content-type') ?? '' })
    }

    clone() {
        if (this.bodyUsed) {
            throw new TypeError('Cannot clone a Response with disturbed body')
        }

        const clonedId = ops.op_response_clone(this[kResponseId])
        const snapshot = ops.op_response_snapshot(clonedId)
        return createResponseFromSnapshot(clonedId, snapshot)
    }

    static error() {
        const response = Object.create(Response.prototype)
        response._status = 0
        response._statusText = ''
        response._type = 'error'
        response._url = ''
        response[kHeaders] = new Headers()
        response[kBodyUsed] = true
        response._contentType = ''
        response[kResponseId] = ''
        return response
    }

    static redirect(url, status = 302) {
        if (!VALID_REDIRECT_STATUSES.has(status)) {
            throw new RangeError('Invalid status code for redirect')
        }

        return new Response(null, {
            status,
            headers: { Location: String(url) },
        })
    }
}

export { Response }

File: lib/web/response.go
package web

import (
	"fmt"
	"sync"

	"github.com/comoland/como/js"
)

const (
	responseTypeBasic              = "basic"
	responseTypeCors               = "cors"
	responseTypeOpaque             = "opaque"
	responseTypeOpaqueRedirect     = "opaqueredirect"
	responseTypeDefault            = responseTypeBasic
	responseTypeError              = "error"
	responseBodyKindNone           = "none"
	responseBodyKindBytes          = "bytes"
	responseBodyKindBlob           = "blob"
	maxResponseStatus          int = 599
	minResponseStatus          int = 200
)

type responseRecord struct {
	id          string
	status      int
	statusText  string
	headers     map[string]string
	url         string
	respType    string
	bodyKind    string
	body        []byte
	blobParts   []string
	contentType string
	bodyUsed    bool
}

func (r *responseRecord) cloneHeaders() map[string]string {
	if len(r.headers) == 0 {
		return map[string]string{}
	}
	cloned := make(map[string]string, len(r.headers))
	for k, v := range r.headers {
		cloned[k] = v
	}
	return cloned
}

func (r *responseRecord) hasBody() bool {
	return r.bodyKind != responseBodyKindNone
}

func (r *responseRecord) cloneWithoutID() *responseRecord {
	var bodyCopy []byte
	if len(r.body) > 0 {
		bodyCopy = make([]byte, len(r.body))
		copy(bodyCopy, r.body)
	}

	var blobPartsCopy []string
	if len(r.blobParts) > 0 {
		blobPartsCopy = make([]string, len(r.blobParts))
		copy(blobPartsCopy, r.blobParts)
	}

	return &responseRecord{
		status:      r.status,
		statusText:  r.statusText,
		headers:     r.cloneHeaders(),
		url:         r.url,
		respType:    r.respType,
		bodyKind:    r.bodyKind,
		body:        bodyCopy,
		blobParts:   blobPartsCopy,
		contentType: r.contentType,
		bodyUsed:    false,
	}
}

type responseStore struct {
	mu      sync.RWMutex
	records map[string]*responseRecord
}

func newResponseStore() *responseStore {
	return &responseStore{
		records: make(map[string]*responseRecord),
	}
}

func (s *responseStore) insert(record *responseRecord) string {
	s.mu.Lock()
	defer s.mu.Unlock()

	record.id = generateUUID()
	s.records[record.id] = record
	return record.id
}

func (s *responseStore) snapshot(id string) (*responseRecord, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	record, ok := s.records[id]
	if !ok {
		return nil, fmt.Errorf("response not found: %s", id)
	}

	copyRecord := record.cloneWithoutID()
	copyRecord.id = record.id
	copyRecord.bodyUsed = record.bodyUsed
	return copyRecord, nil
}

func (s *responseStore) clone(id string) (*responseRecord, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	record, ok := s.records[id]
	if !ok {
		return nil, fmt.Errorf("response not found: %s", id)
	}

	if record.bodyUsed {
		return nil, fmt.Errorf("cannot clone a disturbed Response")
	}

	cloned := record.cloneWithoutID()
	cloned.id = generateUUID()
	s.records[cloned.id] = cloned
	return cloned, nil
}

func (s *responseStore) consumeBody(id string) ([]byte, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	record, ok := s.records[id]
	if !ok {
		return nil, fmt.Errorf("response not found: %s", id)
	}

	if record.bodyUsed {
		return nil, fmt.Errorf("body already used")
	}

	data, err := record.readBody()
	if err != nil {
		return nil, err
	}

	record.bodyUsed = true
	if len(data) > 0 {
		record.bodyKind = responseBodyKindBytes
		record.body = make([]byte, len(data))
		copy(record.body, data)
		record.blobParts = nil
	} else {
		record.body = nil
		record.blobParts = nil
		record.bodyKind = responseBodyKindNone
	}

	return data, nil
}

func (r *responseRecord) readBody() ([]byte, error) {
	switch r.bodyKind {
	case responseBodyKindNone:
		return []byte{}, nil
	case responseBodyKindBytes:
		if len(r.body) == 0 {
			return []byte{}, nil
		}
		data := make([]byte, len(r.body))
		copy(data, r.body)
		return data, nil
	case responseBodyKindBlob:
		if len(r.blobParts) == 0 {
			return []byte{}, nil
		}

		var result []byte
		for _, partID := range r.blobParts {
			part, err := globalBlobStore.GetPart(partID)
			if err != nil {
				return nil, err
			}

			chunk, err := part.Read()
			if err != nil {
				return nil, err
			}
			result = append(result, chunk...)
		}
		return result, nil
	default:
		return nil, fmt.Errorf("unknown body kind: %s", r.bodyKind)
	}
}

type responseCreatePayload struct {
	Status       int               `mapstructure:"status"`
	StatusText   string            `mapstructure:"statusText"`
	Headers      map[string]string `mapstructure:"headers"`
	URL          string            `mapstructure:"url"`
	ResponseType string            `mapstructure:"type"`
	BodyKind     string            `mapstructure:"bodyKind"`
	BlobParts    []string          `mapstructure:"blobParts"`
	ContentType  string            `mapstructure:"contentType"`
}

var globalResponseStore = newResponseStore()

func registerResponse(ctx *js.Context) {
	mod := ctx.NewModule("response.go")
	store := globalResponseStore

	mod.Export("op_response_create", func(args js.Arguments) interface{} {
		if args.Len() == 0 {
			return ctx.Throw("op_response_create requires payload")
		}

		payload := responseCreatePayload{}
		if err := args.GetMap(0, &payload); err != nil {
			return ctx.Throw(err.Error())
		}

		if payload.Status == 0 {
			payload.Status = 200
		}

		if payload.Status < 0 || payload.Status > maxResponseStatus {
			return ctx.Throw(fmt.Sprintf("invalid Response status %d", payload.Status))
		}

		if payload.Headers == nil {
			payload.Headers = map[string]string{}
		}

		respType := payload.ResponseType
		if respType == "" {
			respType = responseTypeDefault
		}

		bodyKind := payload.BodyKind
		if bodyKind == "" {
			bodyKind = responseBodyKindNone
		}

		var bodyBytes []byte
		switch bodyKind {
		case responseBodyKindNone:
			// no-op
		case responseBodyKindBytes:
			if args.Len() < 2 {
				return ctx.Throw("body bytes are required for bodyKind 'bytes'")
			}
			raw, err := args.GetTypedArray(1)
			if err != nil {
				return ctx.Throw("body must be Uint8Array")
			}
			bodyBytes = make([]byte, len(raw))
			copy(bodyBytes, raw)
		case responseBodyKindBlob:
			if len(payload.BlobParts) == 0 {
				return ctx.Throw("blob body requires part IDs")
			}
		default:
			return ctx.Throw(fmt.Sprintf("unsupported bodyKind '%s'", bodyKind))
		}

		record := &responseRecord{
			status:      payload.Status,
			statusText:  payload.StatusText,
			headers:     payload.Headers,
			url:         payload.URL,
			respType:    respType,
			bodyKind:    bodyKind,
			body:        bodyBytes,
			blobParts:   payload.BlobParts,
			contentType: payload.ContentType,
			bodyUsed:    false,
		}

		return store.insert(record)
	})

	mod.Export("op_response_snapshot", func(args js.Arguments) interface{} {
		if args.Len() == 0 {
			return ctx.Throw("op_response_snapshot requires response id")
		}

		id, ok := args.Get(0).(string)
		if !ok {
			return ctx.Throw("response id must be a string")
		}

		record, err := store.snapshot(id)
		if err != nil {
			return ctx.Throw(err.Error())
		}

		return map[string]interface{}{
			"id":          record.id,
			"status":      record.status,
			"statusText":  record.statusText,
			"headers":     record.headers,
			"url":         record.url,
			"type":        record.respType,
			"bodyUsed":    record.bodyUsed,
			"hasBody":     record.hasBody(),
			"contentType": record.contentType,
		}
	})

	mod.Export("op_response_clone", func(args js.Arguments) interface{} {
		if args.Len() == 0 {
			return ctx.Throw("op_response_clone requires response id")
		}

		id, ok := args.Get(0).(string)
		if !ok {
			return ctx.Throw("response id must be string")
		}

		cloned, err := store.clone(id)
		if err != nil {
			return ctx.Throw(err.Error())
		}

		return cloned.id
	})

	mod.Export("op_response_consume_body", func(args js.Arguments) interface{} {
		if args.Len() == 0 {
			return ctx.Throw("op_response_consume_body requires response id")
		}

		id, ok := args.Get(0).(string)
		if !ok {
			return ctx.Throw("response id must be string")
		}

		data, err := store.consumeBody(id)
		if err != nil {
			return ctx.Throw(err.Error())
		}

		return data
	})
}
